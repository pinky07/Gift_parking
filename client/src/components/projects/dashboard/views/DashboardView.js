import React from 'react';
import Toast from 'grommet/components/Toast';

import DashboardWithCriticalError from './viewComponents/DashboardWithCriticalError';
import LoadingView from './viewComponents/LoadingView';
import DashboardWithNoCycleSnaps from './viewComponents/DashboardWithNoCycleSnaps';
import DashboardWithCycleSnaps from './viewComponents/DashboardWithCycleSnaps';

import AddCycleSnapController from '../../addCycleSnap/AddCycleSnapController';
import AddReleaseSnapController from "../../addReleaseSnap/AddReleaseSnapController";
import AddIncidentsReportController from "../../addIncidentsReport/AddIncidentsReportController";

const DashboardView = ({viewModel}) => {
  const {errorMessage} = viewModel;

  if (errorMessage)
    return (<DashboardWithCriticalError errorMessage={errorMessage}/>);
  else {
    const {name} = viewModel.project;
    if (name) {
      const elements = determineElements(viewModel);
      const {cycleSnaps} = viewModel.project;
      if (cycleSnaps.length > 0) {
        return (<DashboardWithCycleSnaps viewModel={viewModel} elements={elements}/>);
      }
      else
        return (<DashboardWithNoCycleSnaps viewModel={viewModel} elements={elements}/>);
    }
    else
      return (<LoadingView/>);
  }
}

/* Determines the UI elements that will be shown: 
  - success notification after adding a cycle snap 
  - failure notification when adding a cycle snap 
  - the modal dialog for adding a cycle snap 
*/
const determineElements = (viewModel) => {
  let elements = {
    successNotification: undefined,
    failureNotification: undefined,
    addCycleSnapLayer: undefined,
    addReleaseSnapLayer: undefined,
    addIncidentsReportLayer: undefined
  };

  const {successNotificationOnAdd} = viewModel;
  if (successNotificationOnAdd)
    elements.successNotification = <Toast status='ok'>{successNotificationOnAdd}</Toast>;

  const {failureNotificationOnAdd} = viewModel;
  if (failureNotificationOnAdd)
    elements.failureNotification = <Toast status='critical'>{failureNotificationOnAdd}</Toast>;

  const {dashboardCallbacks} = viewModel;
  const {addCycleSnap} = viewModel;
  const {addReleaseSnap} = viewModel;
  const {addIncidentsReport} = viewModel;
  
  if (addCycleSnap) {
    elements.addCycleSnapLayer = <AddCycleSnapController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddCycleSnapSubmit}/>;
  }
  
  if (addReleaseSnap) {
    elements.addReleaseSnapLayer = <AddReleaseSnapController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddReleaseSnapSubmit}/>;
  }

  if (addIncidentsReport) {
    elements.addIncidentsReportLayer = <AddIncidentsReportController
      projectId={viewModel.projectId}
      projectName={viewModel.project.name}
      onClose={dashboardCallbacks.onClose}
      onSubmit={dashboardCallbacks.onAddIncidentsReportSubmit}/>;
  }

  return elements;
}

export default DashboardView;