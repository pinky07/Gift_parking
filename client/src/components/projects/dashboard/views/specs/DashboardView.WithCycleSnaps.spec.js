import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import DashboardView from '../DashboardView';

describe('DashboardView when project has one or more cycle snaps', () => {

  let viewModel;
  let actual;

  it('should render dashboard with two cycle snaps', () => {
    viewModel = getViewModelWithTwoCycleSnaps();
    const renderer = ShallowRenderer.createRenderer();
    actual = renderer.render(<DashboardView viewModel={viewModel} />);

    expect(actual).toMatchSnapshot();
  });

  function getViewModelWithTwoCycleSnaps() {
    return {
      projectId: '7',
      project: getProjectWithCycleSnaps(),
      dashboardCallbacks: getDashboardCallbacks(),

      errorMessage: '',
      addCycleSnap: false,
      successNotificationOnAdd: undefined,
      failureNotificationOnAdd: undefined
    }
  }

  function getDashboardCallbacks() {
    const onRequestAddCycleSnap = jest.fn();
    const onAddCycleSnapSubmit = jest.fn();
    const onAddCycleSnapCancel = jest.fn();

    return {
      onRequestAddCycleSnap: onRequestAddCycleSnap,
      onAddCycleSnapSubmit: onAddCycleSnapSubmit,
      onClose: onAddCycleSnapCancel
    }
  }

  function getProjectWithCycleSnaps() {
    // Test data from http://localhost:8080/api/v1/projects/7/dashboard
    return {
      "name": "Exceptional Project",
      "cycleSnaps": [{
        "cycleSnapName": "5",
        "startDate": "2017-03-16",
        "endDate": "2017-03-17",
        "targetedPoints": 1,
        "achievedPoints": 2,
        "tac": "200%",
        "relatedIncidents": "7",
        "daysSinceLastRelease": "15",
        "lastReleaseName": "Second Update",
        "lastReleaseDate": "2017-03-02",
        "teamCapacity": 0.0,
        "wasteDays": 0.0,
        "wastePercentage": "No data",
        "mood": "No data"
      }, {
        "cycleSnapName": "Sprint #4",
        "startDate": "2017-03-01",
        "endDate": "2017-03-15",
        "targetedPoints": 100,
        "achievedPoints": 100,
        "tac": "100%",
        "relatedIncidents": "6",
        "daysSinceLastRelease": "13",
        "lastReleaseName": "Second Update",
        "lastReleaseDate": "2017-03-02",
        "teamCapacity": 0.0,
        "wasteDays": 0.0,
        "wastePercentage": "No data",
        "mood": "No data"
      }, {
        "cycleSnapName": "Sprint #3",
        "startDate": "2017-02-18",
        "endDate": "2017-02-28",
        "targetedPoints": 70,
        "achievedPoints": 68,
        "tac": "97%",
        "relatedIncidents": "6",
        "daysSinceLastRelease": "21",
        "lastReleaseName": "First Version",
        "lastReleaseDate": "2017-02-07",
        "teamCapacity": 99.0,
        "wasteDays": 0.0,
        "wastePercentage": "0.0%",
        "mood": "2.50"
      }, {
        "cycleSnapName": "Sprint #2",
        "startDate": "2017-02-01",
        "endDate": "2017-02-14",
        "targetedPoints": 60,
        "achievedPoints": 55,
        "tac": "91%",
        "relatedIncidents": "5",
        "daysSinceLastRelease": "7",
        "lastReleaseName": "First Version",
        "lastReleaseDate": "2017-02-07",
        "teamCapacity": 99.0,
        "wasteDays": 5.67,
        "wastePercentage": "5.7%",
        "mood": "3.00"
      }, {
        "cycleSnapName": "Sprint #1",
        "startDate": "2017-01-25",
        "endDate": "2017-01-30",
        "targetedPoints": 0,
        "achievedPoints": 0,
        "tac": "No Data",
        "relatedIncidents": "No releases yet",
        "daysSinceLastRelease": "5",
        "lastReleaseName": "No releases yet",
        "lastReleaseDate": "No releases yet",
        "teamCapacity": 100.0,
        "wasteDays": 5.0,
        "wastePercentage": "5.0%",
        "mood": "3.00"
      }]
    };
  }
});