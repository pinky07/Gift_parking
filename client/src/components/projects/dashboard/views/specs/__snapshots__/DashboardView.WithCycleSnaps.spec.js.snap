// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardView when project has one or more cycle snaps should render dashboard with two cycle snaps 1`] = `
<DashboardWithCycleSnaps
  elements={
    Object {
      "addCycleSnapLayer": undefined,
      "addIncidentsReportLayer": undefined,
      "addReleaseSnapLayer": undefined,
      "failureNotification": undefined,
      "successNotification": undefined,
    }
  }
  viewModel={
    Object {
      "addCycleSnap": false,
      "dashboardCallbacks": Object {
        "onAddCycleSnapSubmit": [Function],
        "onClose": [Function],
        "onRequestAddCycleSnap": [Function],
      },
      "errorMessage": "",
      "failureNotificationOnAdd": undefined,
      "project": Object {
        "cycleSnaps": Array [
          Object {
            "achievedPoints": 2,
            "cycleSnapName": "5",
            "daysSinceLastRelease": "15",
            "endDate": "2017-03-17",
            "lastReleaseDate": "2017-03-02",
            "lastReleaseName": "Second Update",
            "mood": "No data",
            "relatedIncidents": "7",
            "startDate": "2017-03-16",
            "tac": "200%",
            "targetedPoints": 1,
            "teamCapacity": 0,
            "wasteDays": 0,
            "wastePercentage": "No data",
          },
          Object {
            "achievedPoints": 100,
            "cycleSnapName": "Sprint #4",
            "daysSinceLastRelease": "13",
            "endDate": "2017-03-15",
            "lastReleaseDate": "2017-03-02",
            "lastReleaseName": "Second Update",
            "mood": "No data",
            "relatedIncidents": "6",
            "startDate": "2017-03-01",
            "tac": "100%",
            "targetedPoints": 100,
            "teamCapacity": 0,
            "wasteDays": 0,
            "wastePercentage": "No data",
          },
          Object {
            "achievedPoints": 68,
            "cycleSnapName": "Sprint #3",
            "daysSinceLastRelease": "21",
            "endDate": "2017-02-28",
            "lastReleaseDate": "2017-02-07",
            "lastReleaseName": "First Version",
            "mood": "2.50",
            "relatedIncidents": "6",
            "startDate": "2017-02-18",
            "tac": "97%",
            "targetedPoints": 70,
            "teamCapacity": 99,
            "wasteDays": 0,
            "wastePercentage": "0.0%",
          },
          Object {
            "achievedPoints": 55,
            "cycleSnapName": "Sprint #2",
            "daysSinceLastRelease": "7",
            "endDate": "2017-02-14",
            "lastReleaseDate": "2017-02-07",
            "lastReleaseName": "First Version",
            "mood": "3.00",
            "relatedIncidents": "5",
            "startDate": "2017-02-01",
            "tac": "91%",
            "targetedPoints": 60,
            "teamCapacity": 99,
            "wasteDays": 5.67,
            "wastePercentage": "5.7%",
          },
          Object {
            "achievedPoints": 0,
            "cycleSnapName": "Sprint #1",
            "daysSinceLastRelease": "5",
            "endDate": "2017-01-30",
            "lastReleaseDate": "No releases yet",
            "lastReleaseName": "No releases yet",
            "mood": "3.00",
            "relatedIncidents": "No releases yet",
            "startDate": "2017-01-25",
            "tac": "No Data",
            "targetedPoints": 0,
            "teamCapacity": 100,
            "wasteDays": 5,
            "wastePercentage": "5.0%",
          },
        ],
        "name": "Exceptional Project",
      },
      "projectId": "7",
      "successNotificationOnAdd": undefined,
    }
  }
/>
`;
